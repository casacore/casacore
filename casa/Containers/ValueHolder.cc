//# ValueHolder.cc: A holder object for the standard AIPS++ data
//# Copyright (C) 2005
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//#
//# $Id$


#include <casa/Containers/ValueHolder.h>
#include <casa/Containers/Record.h>
#include <casa/Arrays/Array.h>
#include <casa/Arrays/ArrayMath.h>
#include <casa/Utilities/Assert.h>
#include <casa/Exceptions/Error.h>

namespace casa { //# NAMESPACE CASA - BEGIN


ValueHolder::ValueHolder (Bool value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (uChar value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (Short value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (uShort value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (Int value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (uInt value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (Float value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (Double value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const Complex& value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const DComplex& value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const Char* value)
  : itsRep (new ValueHolderRep(String(value)))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const String& value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const Array<Bool>& value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const Array<uChar>& value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const Array<Short>& value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const Array<uShort>& value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const Array<Int>& value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const Array<uInt>& value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const Array<Float>& value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const Array<Double>& value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const Array<Complex>& value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const Array<DComplex>& value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const Array<String>& value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const Record& value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (uInt ndim, Bool dummy)
  : itsRep (new ValueHolderRep(ndim, dummy))
{
  itsRep->link();
}


ValueHolder::ValueHolder (const ValueHolder& that)
  : itsRep (that.itsRep)
{
  if (itsRep) itsRep->link();
}

ValueHolder& ValueHolder::operator= (const ValueHolder& that)
{
  if (this != &that) {
    ValueHolderRep::unlink (itsRep);
    itsRep = that.itsRep;
    if (itsRep) itsRep->link();
  }
  return *this;
}


} //# NAMESPACE CASA - END
