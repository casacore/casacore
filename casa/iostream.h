//# iostream.h: Interim solution for standard/nonstandard system iostream
//# Copyright (C) 2001,2002
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef CASA_IOSTREAM_H
#define CASA_IOSTREAM_H

//# Define the C standard C++ include file. 
//# This is an interim solution to cater for the SGI non-existence of
//# them (e.g. <cstring>)

//# Make sure any special macros are set
#include <casacore/casa/aips.h>
#include <iostream>

namespace casacore { //# NAMESPACE CASACORE - BEGIN

using std::istream;
using std::ostream;
using std::iostream;
using std::cin;
using std::cout;
using std::cerr;
using std::endl;
using std::flush;
using std::ws;

using std::ios;
using std::streampos;
using std::streamoff;
using std::streamsize;
using std::dec;
using std::hex;
using std::oct;

using std::internal;
using std::left;
using std::right;
using std::fixed;
using std::scientific;

using std::boolalpha;
using std::noboolalpha;
using std::showbase;
using std::noshowbase;
using std::showpoint;
using std::noshowpoint;
using std::showpos;
using std::noshowpos;
using std::skipws;
using std::noskipws;
using std::uppercase;
using std::nouppercase;

} //# NAMESPACE CASACORE - END

#endif
