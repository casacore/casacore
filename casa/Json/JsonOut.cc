//# JsonOut.cc: Fill a file or stream in Json format
//# Copyright (C) 2016
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$


//# Includes
#include <casacore/casa/Json/JsonOut.h>
#include <casacore/casa/Json/JsonError.h>
#include <casacore/casa/BasicSL/String.h>
#include <casacore/casa/Arrays/Array.h>
#include <casacore/casa/Containers/Record.h>
#include <casacore/casa/Containers/ValueHolder.h>
#include <iomanip>

namespace casacore { //# NAMESPACE CASACORE - BEGIN

  JsonOut::JsonOut()
    : itsStream (cout),
      itsLevel  (0)
  {}

  JsonOut::JsonOut (const String& name)
    : itsFile   (name.chars()),
      itsStream (itsFile),
      itsLevel  (0)
  {}

  JsonOut::JsonOut (ostream& os)
    : itsStream (os),
      itsLevel  (0)
  {}

  // Close the stream.
  JsonOut::~JsonOut()
  {}

  void JsonOut::start (const String& commentStart, const String& commentEnd,
                       const String& indent)
  {
    AlwaysAssert (itsLevel==0, JsonError);
    itsStream << '{' << endl;
    itsIndent       = indent;
    itsIndentStep   = indent;
    itsCommentStart = commentStart;
    itsCommentEnd   = commentEnd;
    itsLevel = 1;
    itsFirstName.resize (1);
    itsFirstName[0] = True;
  }

  void JsonOut::end()
  {
    itsLevel--;
    AlwaysAssert (itsLevel==0, JsonError);
    itsIndent.clear();
    itsStream << '}' << endl;
  }

  void JsonOut::startNested (const String& name, const String& comment)
  {
    AlwaysAssert (itsLevel>0, JsonError);
    writeComment (comment);
    putName (name);
    itsStream << '{' << endl;
    itsIndent += itsIndentStep;
    itsLevel++;
    itsFirstName.resize (itsLevel);
    itsFirstName[itsLevel-1] = True;
  }

  void JsonOut::endNested()
  {
    itsLevel--;
    AlwaysAssert (itsLevel>0, JsonError);
    itsIndent = itsIndent.substr (0, itsIndent.size() - itsIndentStep.size());
    itsStream << itsIndent << '}' << endl;
  }

  void JsonOut::writeKV (const String& name, const ValueHolder& vh)
  {
    if (vh.isNull()) {
      putNull();
    } else {
      switch (vh.dataType()) {
      case TpBool:
        writeKV (name, vh.asBool());
        break;
      case TpChar:
      case TpUChar:
      case TpShort:
      case TpUShort:
      case TpInt:
      case TpUInt:
      case TpInt64:
        writeKV (name, vh.asInt64());
        break;
      case TpFloat:
        writeKV (name, vh.asFloat());
        break;
      case TpDouble:
        writeKV (name, vh.asDouble());
        break;
      case TpComplex:
        writeKV (name, vh.asComplex());
        break;
      case TpDComplex:
        writeKV (name, vh.asDComplex());
        break;
      case TpString:
        writeKV (name, vh.asString());
        break;
      case TpArrayBool:
        writeKV (name, vh.asArrayBool());
        break;
      case TpArrayChar:
      case TpArrayUChar:
      case TpArrayShort:
      case TpArrayUShort:
      case TpArrayInt:
      case TpArrayUInt:
      case TpArrayInt64:
        writeKV (name, vh.asArrayInt64());
        break;
      case TpArrayFloat:
        writeKV (name, vh.asArrayFloat());
        break;
      case TpArrayDouble:
        writeKV (name, vh.asArrayDouble());
        break;
      case TpArrayComplex:
        writeKV (name, vh.asArrayComplex());
        break;
      case TpArrayDComplex:
        writeKV (name, vh.asArrayDComplex());
        break;
      case TpArrayString:
        writeKV (name, vh.asArrayString());
        break;
      case TpRecord:
        writeKV (name, vh.asRecord());
        break;
      default:
        throw JsonError("JsonOut: unknown ValueHolder datatype");
      }
    }
  }

  void JsonOut::writeComment (const String& comment)
  {
    if (!itsCommentStart.empty()  &&  !comment.empty()) {
      itsStream << itsIndent << ' ' << itsCommentStart << ' ' << comment;
      if (!itsCommentEnd.empty()) {
        itsStream << itsCommentEnd;
      }
      itsStream << endl;
    } 
 }

  String JsonOut::indentValue (const String& indent, const String& name) const
  {
    String extra("                    ");
    return indent + extra.substr(0, std::min(15,int(name.size())) + 5);
  }

  void JsonOut::putName (const String& name)
  {
    itsStream << itsIndent;
    if (itsFirstName[itsLevel-1]) {
      itsStream << ' ';
      itsFirstName[itsLevel-1] = False;
    } else {
      itsStream << ',';
    }
    itsStream << '"' << name << "\": ";
  }

  void JsonOut::putNull()
  {
    itsStream << "null";
  }

  void JsonOut::put (Bool value)
  {
    itsStream << (value ? "true" : "false");
  }
  void JsonOut::put (Float value)
  {
    if (! isFinite(value)) {
      putNull();
    } else {
      char buf[16];
      sprintf (buf, "%.7g", value);
      itsStream << buf;
    }
  }
  void JsonOut::put (Double value)
  {
    if (! isFinite(value)) {
      putNull();
    } else {
      char buf[24];
      sprintf (buf, "%.16g", value);
      itsStream << buf;
    }
  }
  void JsonOut::put (const Complex& value)
  {
    itsStream << "{\"r\":";
    put (value.real());
    itsStream << ", \"i\":";
    put (value.imag());
    itsStream << '}';
  }
  void JsonOut::put (const DComplex& value)
  {
    itsStream << "{\"r\":";
    put (value.real());
    itsStream << ", \"i\":";
    put (value.imag());
    itsStream << '}';
  }
  void JsonOut::put (const char* value)
    { itsStream << '"' << value << '"'; }
  void JsonOut::put (const String& value)
    { itsStream << '"' << value << '"'; }

  void JsonOut::put (const Record& rec)
  {
    itsStream << '{' << endl;
    String oldIndent(itsIndent);
    itsIndent += itsIndentStep;
    itsLevel++;
    itsFirstName.resize (itsLevel);
    itsFirstName[itsLevel-1] = True;
    for (uInt i=0; i<rec.nfields(); ++i) {
      write (rec.name(i), rec.asValueHolder(i));
    }
    itsLevel--;
    itsIndent = oldIndent;
    itsStream << itsIndent << '}';
  }

  String JsonOut::escapeString (const String& in)
  {
    String out;
    out.reserve (in.size());
    for (size_t i=0; i<in.size(); ++i) {
      if (in[i]== '"'  ||  in[i] == '\\') {
        out.append ('\\');
      }
      out.append (in[i]);
    }
    return out;
  }

} //# NAMESPACE CASACORE - END
