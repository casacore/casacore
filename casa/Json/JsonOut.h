//# JsonOut.h: Fill a file or stream in JSON format
//# Copyright (C) 2016
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef CASA_JSONOUT_H
#define CASA_JSONOUT_H

//# Includes
#include <casacore/casa/aips.h>
#include <casacore/casa/BasicSL/String.h>
#include <casacore/casa/Arrays/Array.h>
#include <iostream>
#include <fstream>

namespace casacore { //# NAMESPACE CASACORE - BEGIN

  //# Forward Declarations
  class Record;
  class ValueHolder;


  // <summary>
  // Class to fill a file or stream in JSON format.
  // </summary>

  // <use visibility=export>
  // <reviewed reviewer="" date="" tests="tJsonOut">
  // </reviewed>

  //# <prerequisite>
  //# </prerequisite>

  // <synopsis>
  // JsonOut is a class to create a JSON file. JsonParser.h can be used
  // to interpret a JSON file whereafter JsonKVMap gets out the information.
  //
  // Besides the standard JSON types (bool, int, float, string), sequences
  // and nested structs, JsonOut also supports Casacore data type (D)Complex,
  // Array, Record, and ValueHolder.
  // <br>- A complex number is written as a nested struct with fields
  //       "r" and "i".
  // <br>- An Array is written as a (possibly nested) sequence of values.
  // <br>- A Record is written as a nested struct; subrecords are supported.
  // <br>- A ValueHolder is written depending on the data type it contains.
  // <br>Note that floating point values are written with high accuracy
  // (7 digits for single precision, 16 digits for double precision).
  //
  // Although standard JSON does not support comments, many parsers do support
  // C-style and C++-style comments. JsonOut has the possibility to define
  // arbitrary comment delimiters (e.g., / * and * / for C-style).
  // If no start delimiter is given, possible comments are ignored.
  //
  // The output of JsonOut can be any iostream. If a file name is given, an
  // ofstream will be opened in the constructor and closed in the destructor.
  // The output is formatted pretty nicely. Nested structs are indented with
  // 2 spaces. Arrays are written with a single axis per line; continuation
  // lines are indented properly. String arrays have one value per line.
  // </synopsis>

  // <example>
  // The following example is read back by the example in class JsonParser.
  // <srcblock>
  // // Create the JSON file.
  // JsonOut jout(fullName + "/imageconcat.json");
  // // Start the JSON struct; possible comments will be ignored.
  // jout.start();
  // // Write some fields (one line per field).
  // jout.write ("Version", 1);
  // jout.write ("DataType", "float");
  // jout.write ("Axis", latticeConcat_p.axis());
  // jout.write ("Images", Array<String>(latticeNames));
  // // End the JSON struct.
  // jout.end();
  // </srcblock>
  // See tJsonOut.cc for more elaborate examples.
  // </example>

  // <motivation>
  // JSON is a commonly used interchange format.
  // </motivation>
  //
  //# <todo asof="1996/03/10">
  //#   <li> 
  //# </todo>

  class JsonOut
  {
  public:
    // The default constructor creates the output on stdout.
    JsonOut();

    // Create the file with the given name using an ofstream object.
    JsonOut (const String& name);

    // Create the object using the given ostream object.
    JsonOut (ostream& os);

    // Close the stream. It closes the ofstream object if created.
    ~JsonOut();

    // Start a JSON structure by writing a { and setting the indentation.
    // It checks if not inside a JSON structure.
    // It is possible to define the comment delimiters
    // (e.g., / * and * /  or  // and empty).
    // If commentStart is empty, possible comments are ignored.
    void start (const String& commentStart=String(),
                const String& commentEnd=String(),
                const String& indent="  ");

    // End a structure by clearing the indentation and writing a }.
    // It checks if inside a JSON structure.
    void end();

    // Start a nested structure; i.e., a field with a structured value.
    // It writes the name and opening brace and increments the indentation.
    // If supported, the comment is written on a line preceeding the key line.
    void startNested (const String& name, const String& comment=String());

    // End a nested structure.
    // It decrements the indentation and writes the closing brace.
    void endNested();

    // Write one or more lines defining a keyword-value pair, where value
    // can be of any type including Array, Record, and ValueHolder.
    // A non-finite floating point number and a null ValueHolder are
    // written as a null value.
    // If supported, the comment is written on a line preceeding the
    // 'key:value' line.
    template <typename T>
    void write (const String& name, T value, const String& comment=String());

    // Write a comment on a separate line.
    // If comments are not supported, an empty line is written.
    void writeComment (const String& comment);

    // Write a null value.
    void putNull();

    // Put a scalar value with sufficient accuracy.
    // A Complex value is written as a nested JSON structure
    // with fields r and i.
    // A string is enclosed in quotes and escaped where necessary.
    // <br>These functions are meant for internal use by the 'write' function.
    // <group>
    template <typename T> void put (T value);
    void put (Bool value);
    void put (Float value);
    void put (Double value);
    void put (const Complex& value);
    void put (const DComplex& value);
    void put (const char* value);
    void put (const String& value);
    // </group>

    // Put a line defining an array value. Multi-dim arrays are written as
    // nested [] lines.
    // Normally the values of the first dimension are written on a single line,
    // but for string values a line per value is used.
    // <br>These functions are meant for internal use by the 'write' function.
    template <typename T>
    void putArray (const Array<T>& value, const String& indent,
                   Bool firstLine);
    void putArray (const Array<String>& value, const String& indent,
                   Bool firstLine);
    template <typename T>
    void putArray (const Array<T>& value, const String& indent,
                   Bool firstLine, Bool valueEndl);

    // Escape a double quote and backslash in a string.
    static String escapeString (const String& in);

  private:
    // Copy constructor cannot be used.
    JsonOut (const JsonOut& other);

    // Assignment cannot be used.
    JsonOut& operator= (const JsonOut& other);

    // Write the name.
    void putName (const String& name);

    // General function to write a key and value.
    // Specializations exist for particular data types.
    template <typename T>
    void writeKV (const String& name, T value);

    // Write a key and array value.
    template <typename T>
    void writeKV (const String& name, const Array<T>& value);

    // Write a key and valueholder.
    void writeKV (const String& name, const ValueHolder& vh);

    // Put a Record which is written as a {} structure.
    // The Record can be nested.
    void put (const Record&);

    // Get the indentation after a name.
    // It indents with the length of the name (including quotes and colon)
    // with a maximum of 20 spaces.
    String indentValue (const String& indent, const String& name) const;

    //# Data fields.
    std::ofstream itsFile;
    std::ostream& itsStream;
    String        itsIndent;
    String        itsIndentStep;
    int           itsLevel;
    String        itsCommentStart;
    String        itsCommentEnd;
    vector<Bool>  itsFirstName;
  };


} //# NAMESPACE CASACORE - END

#ifndef CASACORE_NO_AUTO_TEMPLATES
#include <casacore/casa/Json/JsonOut.tcc>
#endif //# CASACORE_NO_AUTO_TEMPLATES
#endif
