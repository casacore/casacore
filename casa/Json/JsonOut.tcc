//# JsonOut.tcc: Fill a file or stream in Json format
//# Copyright (C) 2016
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$


//# Includes
#include <casacore/casa/Json/JsonOut.h>
#include <casacore/casa/BasicSL/String.h>
#include <casacore/casa/Arrays/Array.h>
#include <casacore/casa/Containers/Record.h>

namespace casacore { //# NAMESPACE CASACORE - BEGIN

  template <typename T>
  inline void JsonOut::write (const String& name, T value,
                              const String& comment)
  {
    writeComment (comment);
    putName (name);
    writeKV (name, value);
    itsStream << endl;
  }

  template <typename T>
  inline void JsonOut::writeKV (const String&, T value)
  {
    put (value);
  }

  template <typename T>
  inline void JsonOut::writeKV (const String& name, const Array<T>& value)
  {
    // Use extra indentation for possible continuation lines.
    putArray (value, indentValue(itsIndent, name), True);
  }

  template <typename T>
  inline void JsonOut::put (T value)
    { itsStream << value; }

  template <typename T>
  inline void JsonOut::putArray (const Array<T>& arr,
                                 const String& indent, Bool firstLine)
  {
    putArray (arr, indent, firstLine, False);
  }
  inline void JsonOut::putArray (const Array<String>& arr,
                                 const String& indent, Bool firstLine)
  {
    putArray (arr, indent, firstLine, True);
  }

  template <typename T>
  void JsonOut::putArray (const Array<T>& arr, const String& indent,
                          Bool firstLine, Bool valueEndl)
  {
    if (!firstLine) itsStream << indent;
    itsStream << '[';
    Bool first = True;
    if (arr.ndim() <= 1) {
      size_t todo = arr.size();
      typename Array<T>::const_iterator iterEnd = arr.end();
      for (typename Array<T>::const_iterator iter=arr.begin();
           iter!=iterEnd; ++iter) {
        if (first) {
          first = False;
        } else if (!valueEndl) {
          itsStream << ", ";
        } else {
          itsStream << indent << ' ';
        }
        put (*iter);
        todo--;
        if (valueEndl  &&  todo > 0) {
          itsStream << ',' << endl;
        }
      }
    } else {
      ArrayIterator<T> iter(arr, IPosition(1, arr.ndim()-1), False);
      while (! iter.pastEnd()) {
        if (!first) {
          itsStream << ',' << endl;
        }
        putArray (iter.array(), indent+' ', first);
        first = False;
        iter.next();
      }
    }
    itsStream << ']';
  }


} //# NAMESPACE CASACORE - END
