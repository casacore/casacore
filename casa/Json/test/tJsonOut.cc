//# tJsonOut.cc: Test program for class JsonOut.cc
//# Copyright (C) 2016
//# Associated Universities, Inc. Washington DC, USA.
//# 
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//# 
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//# 
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//# 
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <casacore/casa/Json/JsonOut.h>
#include <casacore/casa/Arrays/ArrayMath.h>
#include <casacore/casa/Containers/ValueHolder.h>
#include <iostream>
#include <stdexcept>

using namespace casacore;

int main()
{
  try {
    String indent("     ");
    JsonOut os("tJsonOut_tmp.txt");
    os.start("#");
    for (int ndim=0; ndim<4; ++ndim) {
      Array<Int> a(IPosition(ndim,3));
      indgen(a);
      os.write ("arr", a);
    }
    Array<String> as(IPosition(2,2,2));
    as.data()[0] = "s0";
    as.data()[1] = "s1";
    as.data()[2] = "s2";
    as.data()[3] = "s3";
    os.write ("arrs", as, "comment1");
    os.write ("key1", true);
    os.write ("key2", 1, "comment2");
    os.write ("key3", -10);
    os.write ("key4", 11.5f);
    os.write ("key5", -13.2345);
    os.write ("key6", Complex(-1,2));
    os.write ("key7", ValueHolder(DComplex(3,-4)));
    os.write ("key8", "string");
    os.write ("null1", ValueHolder());     // null value
    os.write ("null2", floatNaN());        // null value
    os.write ("null3", doubleNaN());       // null value
    Array<Float> arr(IPosition(2,5,3));
    indgen(arr);
    os.write ("arrf", ValueHolder(arr));
    Record rec;
    rec.define ("recfld", 1);
    rec.define ("recarr", arr);
    Record subrec;
    subrec.define("subf1", 2);
    subrec.define("subf2", arr+float(15));
    rec.defineRecord ("recsub", subrec);
    os.write ("rec", rec);
    os.end();
  } catch (std::exception& x) {
    std::cout << x.what() << std::endl;
    return 1;
  }
  return 0;
}
