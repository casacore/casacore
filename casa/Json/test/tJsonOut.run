#! /bin/sh
#-----------------------------------------------------------------------------
# Script to execute the tJsonOut program via assay.
#=============================================================================

$casa_checktool ./tJsonOut
cat tJsonOut_tmp.txt

