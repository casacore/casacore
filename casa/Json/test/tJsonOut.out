{
   "arr": []
  ,"arr": [0, 1, 2]
  ,"arr": [[0, 1, 2],
           [3, 4, 5],
           [6, 7, 8]]
  ,"arr": [[[0, 1, 2],
            [3, 4, 5],
            [6, 7, 8]],
           [[9, 10, 11],
            [12, 13, 14],
            [15, 16, 17]],
           [[18, 19, 20],
            [21, 22, 23],
            [24, 25, 26]]]
   # comment1
  ,"arrs": [["s0",
             "s1"],
            ["s2",
             "s3"]]
  ,"key1": true
   # comment2
  ,"key2": 1
  ,"key3": -10
  ,"key4": 11.5
  ,"key5": -13.2345
  ,"key6": {"r":-1, "i":2}
  ,"key7": {"r":3, "i":-4}
  ,"key8": "string"
  ,"null1": null
  ,"null2": null
  ,"null3": null
  ,"arrf": [[0, 1, 2, 3, 4],
            [5, 6, 7, 8, 9],
            [10, 11, 12, 13, 14]]
  ,"rec": {
     "recfld": 1
    ,"recarr": [[0, 1, 2, 3, 4],
                [5, 6, 7, 8, 9],
                [10, 11, 12, 13, 14]]
    ,"recsub": {
       "subf1": 2
      ,"subf2": [[15, 16, 17, 18, 19],
                 [20, 21, 22, 23, 24],
                 [25, 26, 27, 28, 29]]
    }
  }
}
